### ISSUE TITLE
Using local patches for dev dependencies breaks projects that use cweagans/composer-patches

### ISSUE THREAD

**Issue by Berdir** [predis/predis#660]
**Describe the bug**

Follow-up of https://github.com/predis/predis/issues/632, which didn't fully address this issue 

Adding patches to composer.json for dev dependencies seems incorrect to me, for two reasons:

a) The local patch files won't be found in the module, AFAIK composer-patches won't look there
b) Even if they would be found, the project is likely to use a different version of phpunit where the patches don't apply.

**To Reproduce**
1. composer init
2. composer require cweagans/composer-patches predis/predis
3. composer config extra.enable-patching true
4. composer require --dev phpunit/phpunit

```
  - Applying patches for phpunit/phpunit
    ./tests/phpunit_php7.patch (Fix PHP 7 compatibility)
   Could not apply patch! Skipping. The error was: The "./tests/phpunit_php7.patch" file could not be downloaded: failed to open stream: No such file or directory
    ./tests/phpunit_php8.patch (Fix PHP 8 compatibility)
   Could not apply patch! Skipping. The error was: The "./tests/phpunit_php8.patch" file could not be downloaded: failed to open stream: No such file or directory
```

**Expected behavior**

**Versions (please complete the following information):**
 - Predis: 1.1.x
 - PHP 7.4
 - Redis Server irrelevant
 - OS: Ubuntu 20.10

I don't think it is possible to do this, not without cweagans/composer-patches introducing a patches-dev. My suggestion would be to switch to applying those patches with a script/travis commands (or whatever CI you use)

**Comment by tillkruss** [comment]
The patches are including in the repo now, did you happen to upgrade from 1.1.1 to 1.1.x by chance?

**Comment by Berdir** [comment]
that's my point, even if they would be found, they wouldn't apply to a current version of phpunit.

**Comment by tillkruss** [comment]
The replicated error:

```
Could not apply patch! Skipping. The error was: The "./tests/phpunit_php7.patch" file could not be downloaded: failed to open stream: No such file or directory
```

**Comment by tillkruss** [comment]
@Berdir Have you tried using `patches-ignore` to skip the Predis patches?

**Comment by Berdir** [comment]
I have not, but it shouldn't be the responsibility of every project to work around this?

**Comment by fgm** [comment]
@tillkruss could you suggest the recommended format of `patches-ignore` until a better solution can be found ?

**Comment by tillkruss** [comment]
@fgm: That looks good.

@cweagans: Do you have any suggestions with this?

**Comment by iijiang** [comment]
patches-ignore seems not working on cweagans/composer-patches 1.6.7 @cweagans

**Comment by cweagans** [comment]
Local patches really shouldn't be used by projects that can be pulled in as a dependency in the first place. I recommend either applying those patches in your travis job manually or putting the patches somewhere public where they can be downloaded (maybe Gist? worked pretty well like this: https://github.com/cweagans/mcrypt-polyfill/blob/d261945011d06273fa3221e68cb2b7e75a40b3ee/composer.json#L30).

I'm reasonably sure patches-ignore is working, but if you can reliably reproduce a case where it's not, definitely open an issue with the info and maybe we can get it working.

**Comment by roborourke** [comment]
Ran into this issue as well. There's a `patches-dev` option in this composer plugin https://github.com/vaimo/composer-patches/blob/master/docs/USAGE_BASIC.md that might help.

Also would be good to add the version constraints as well if you don't switch to applying the patches during the travis builds eg:

```json
{
  "extra": {
    "patches-dev": {
      "phpunit/phpunit": {
        "Fix PHP 7 compatibility": {
          "source": "./tests/phpunit_php7.patch",
          "depends": {
            "phpunit/phpunit": "~4.8",
            "php": "7"
          },
          "Fix PHP 8 compatibility": {
            "source": "./tests/phpunit_php8.patch",
            "depends": {
              "phpunit/phpunit": "~4.8",
              "php": "8"
            }
          }
        }
      },
      "phpunit/phpunit-mock-objects": {
        "Fix PHP 7 and 8 compatibility": {
          "source": "./tests/phpunit_mock_objects.patch",
          "depends": {
            "php": ">=7 <9"
          }
       }
    }
  }
}
```

I see 2.x has fixed this issue, do you think this could be patched for v1 as well? I have a situation where applying those patches to the root project's version of PHPUnit would be a problem.

**Comment by Berdir** [comment]
can confirm that patches-ignore  works, but that's not a solution, forcing everyone else to work around this is really confusing.

@cweagans making the patches public won't help here, the patches here are for making old, unsupported versions of phpunit php7+ compatible. They won't apply for anyone who is e.g. using Drupal 8 and most other projects as they are already on newer phpunit versions that don't need those patches.

Having support for patches-dev would work, then local patches are fine. Without that, yes, just add the patches in CI. not sure about the advantages/disadvantages of that other patches package.

@roborourke I'm assume the patches are used because 1.x also supports older php versions which in turn aren't supported by new phpunit versions which is why it uses that hack to test it on both old and new php versions. So it's not "fixed" in v2, it simply doesn't need the patches there as the phpunit version is newer.

**Comment by tillkruss** [comment]
@nrk: Do you have a preference to storing the patches accessible with HTTPS, or patching the files as suggested via Travis?

**Comment by tillkruss** [comment]
@predis/team Anybody up for a PR to apply the patches in Travis itself?

**Comment by nrk** [comment]
@tillkruss anything is fine to me as long as it fixes this problem but I'd prefer a solution that does not require some voodoo magic to have tests up and running locally, aside from CI on Travis... because yeah I do run tests locally on older PHP releases for v1.1.

**Comment by nrk** [comment]
Got extremely busy once again unfortunately so I had no time in the last 6 weeks or so.

So in the end there's no solution for this issue? Maybe an acceptable workaround would be to apply these patches _manually_ (using an utility script at least) instead of relying on composer patches if we want to keep testing `v1.1` test on PHP <= 7.2?

**Comment by tillkruss** [comment]
@nrk: Yes, that's the best and relatively easy solution. It seems that very few people experience this issue.

I've been slammed as well, anyone up for a PR?

**Comment by roborourke** [comment]
If no one has picked it up by then I'll take a stab at it on Monday afternoon.

**Comment by roborourke** [comment]
Bit later than I hoped but I have what I believe is a working PR up at #673. It does require the `patch` utility to be available on the system however so Windows users might need to use WSL or gitbash.

**Comment by smustgrave** [comment]
Will this be fixed in the next predis release?

**Comment by tillkruss** [comment]
@smustgrave: Yes, I'll tag a new release as soon as I get Travis working.

**Comment by tillkruss** [comment]
Closed via #673.
